server/src/main/java/org/opensearch/cluster/routing/allocation/decider/RemoteStoreMigrationAllocationDecider.java

```diff
@@ -0,0 +1,182 @@
+import org.opensearch.cluster.routing.ShardRouting;
+import org.opensearch.cluster.routing.allocation.RoutingAllocation;
+import org.opensearch.common.settings.ClusterSettings;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.node.remotestore.RemoteStoreNodeService;
+import org.opensearch.node.remotestore.RemoteStoreNodeService.CompatibilityMode;
+import org.opensearch.node.remotestore.RemoteStoreNodeService.Direction;
+
+import java.util.Locale;
+
+/**
+ * A new allocation decider for migration of document replication clusters to remote store backed clusters:
+ * - For STRICT compatibility mode, the decision is always YES
+ * - For remote_store_enabled indices, relocation or allocation/relocation can only be towards a remote node
+ * - For "REMOTE_STORE" migration direction:
+ *      - New primary shards can only be allocated to a remote node
+ *      - New replica shards can be allocated to a remote node iff the primary has been migrated/allocated to a remote node
+ * - For other directions ("DOCREP", "NONE"), the decision is always YES
+ *
+ * @opensearch.internal
+ */
+public class RemoteStoreMigrationAllocationDecider extends AllocationDecider {
+
+    public static final String NAME = "remote_store_migration";
+
+    private Direction migrationDirection;
+    private CompatibilityMode compatibilityMode;
+    private boolean remoteStoreEnabledIndex = false;
+
+    public RemoteStoreMigrationAllocationDecider(Settings settings, ClusterSettings clusterSettings) {
+        this.migrationDirection = RemoteStoreNodeService.MIGRATION_DIRECTION_SETTING.get(settings);
+        this.compatibilityMode = RemoteStoreNodeService.REMOTE_STORE_COMPATIBILITY_MODE_SETTING.get(settings);
+        this.remoteStoreEnabledIndex = IndexMetadata.INDEX_REMOTE_STORE_ENABLED_SETTING.get(settings);
+        if (IndexMetadata.INDEX_REMOTE_STORE_ENABLED_SETTING.exists(settings)) {
+            remoteStoreEnabledIndex = IndexMetadata.INDEX_REMOTE_STORE_ENABLED_SETTING.get(settings);
+        }
+        clusterSettings.addSettingsUpdateConsumer(RemoteStoreNodeService.MIGRATION_DIRECTION_SETTING, this::setMigrationDirection);
+        clusterSettings.addSettingsUpdateConsumer(
+            RemoteStoreNodeService.REMOTE_STORE_COMPATIBILITY_MODE_SETTING,
+            this::setCompatibilityMode
+        );
+    }
+
+    // listen for changes in migration direction of cluster
+    private void setMigrationDirection(Direction migrationDirection) {
+        this.migrationDirection = migrationDirection;
+    }
+
+    // listen for changes in compatibility mode of cluster
+    private void setCompatibilityMode(CompatibilityMode compatibilityMode) {
+        this.compatibilityMode = compatibilityMode;
+    }
+
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        DiscoveryNode targetNode = node.node();
+
+        if (compatibilityMode.equals(CompatibilityMode.STRICT)) {
```

Comment:
Lets add the assumption here that all nodes are suppose to be `remote/non-remote` .

Reply:
Added a comment about assumption.

Reply:
Slightly off-topic but we should ideally not allow changing the compatibility mode from `MIXED` to `STRICT` if the cluster already has a mix and match of remote and non-remote nodes.

This would be hard to implement in the Decider or on `RemoteStoreNodeService`, but this can be done on the transport layer during the Cluster settings update request.

Reply:
@ltaragi : lets create a follow up item and resolve this .